### PR TITLE
Add items to collection via context menu

### LINKED ISSUE #37: Add items to collection via context menu
Necessary for accessibility

https://forums.zotero.org/discussion/comment/365900/#Comment_365900

### LINKED ISSUE #2069: Create Collection from Selected Items
https://forums.zotero.org/discussion/89852/right-click-menu-option-create-collection-from-selected-items

I thought we had a ticket for this but I can't find one.

### PR DESCRIPTION
**Pull request by AbeJellinek**
Menu layout modeled off of Advanced Search. We could do submenus, but I can't think of a "File Here" wording that I really like. It's possible to make menu items that contain submenus clickable, but that behavior is unusual.

Closes #37, closes #2069.

### CODE REVIEW

**Comment by dstillman** [comment]
And we could consider having a Move option as well, as an accessible equivalent to [holding down Cmd/Shift](https://www.zotero.org/support/collections_and_tags#adding_items_to_collections).

**Comment by AbeJellinek** [comment]
OK, see new commits. Would Move be a top-level submenu next to Add to Collection?

**Comment by dstillman** [comment]
Yeah, but let's do without for the moment, since it's easy to accomplish the same thing without it. We can revisit if people really want it.

File: chrome/content/zotero/zoteroPane.xul [hunk 1]
```diff
@@ -299,6 +299,12 @@
 							<menuseparator/>
 							<menuitem class="menuitem-iconic zotero-menuitem-toggle-read-item" oncommand="ZoteroPane_Local.toggleSelectedItemsRead();"/>
 							<menuitem class="menuitem-iconic zotero-menuitem-duplicate-item" label="&zotero.items.menu.duplicateItem;" oncommand="ZoteroPane_Local.duplicateSelectedItem().done();"/>
+							<menu class="menuitem-iconic zotero-menuitem-add-to-collection">
+								<menupopup id="zotero-add-to-collection-popup" onpopupshowing="ZoteroPane_Local.buildAddToCollectionMenu(event)">
+									<menuseparator id="zotero-add-to-collection-separator"/>
+									<menuitem id="zotero-add-to-new-collection" label="&zotero.toolbar.newCollection.label;" oncommand="ZoteroPane_Local.addItemsToCollection(null, true)"/>
```
Comment (dstillman) [inline chrome/content/zotero/zoteroPane.xul:305]: Oh, nice — but let's move this above the hierarchical tree. I have a ton of collections in my test library and didn't even notice this existed until I looked at the code

Also can note in the description that this closes https://github.com/zotero/zotero/issues/2069

File: chrome/content/zotero/zoteroPane.js [hunk 2]
```diff
@@ -3080,6 +3090,52 @@ var ZoteroPane = new function()
...
+
+		separator.setAttribute('hidden', !collections.length);
+	};
+
+
+	this.addItemsToCollection = async function (collection, createNew = false) {
+		// Get items first because newCollection() will deselect
+		let items = Zotero.Items.keepParents(this.getSelectedItems());
+
+		if (createNew) {
```
Comment (dstillman) [inline chrome/content/zotero/zoteroPane.js:3130]: Throw error (e.g., "collection must be null if createNew is true") if `collection` is not falsy

File: chrome/content/zotero/zoteroPane.js [hunk 3]
```diff
@@ -3080,6 +3090,52 @@ var ZoteroPane = new function()
...
+				collection => items.every(item => collection.hasItem(item))
+			);
+			separator.before(menuItem);
+		}
+
+		separator.setAttribute('hidden', !collections.length);
+	};
+
+
+	this.addItemsToCollection = async function (collection, createNew = false) {
```
Comment (dstillman) [inline chrome/content/zotero/zoteroPane.js:3126]: → `addSelectedItemsToCollection`

File: chrome/content/zotero/zoteroPane.js [hunk 4]
```diff
@@ -3080,6 +3090,52 @@ var ZoteroPane = new function()
 		// add locate menu options
 		yield Zotero_LocateMenu.buildContextMenu(menu, true);
 	});
+
+
+	this.buildAddToCollectionMenu = function (event) {
+		if (event.target.id !== 'zotero-add-to-collection-popup') return;
+
+		let popup = document.querySelector('#zotero-add-to-collection-popup');
+		let separator = popup.querySelector('#zotero-add-to-collection-separator');
```
Comment (dstillman) [inline chrome/content/zotero/zoteroPane.js:3099]: Might as well use `document.getElementById`

File: chrome/content/zotero/xpcom/utilities_internal.js [hunk 5]
```diff
@@ -1693,16 +1693,19 @@ Zotero.Utilities.Internal = {
 	 * Create a libraryOrCollection DOM tree to place in <menupopup> element.
 	 * If has no children, returns a <menuitem> element, otherwise <menu>.
 	 * 
-	 * @param {Library/Collection} libraryOrCollection
+	 * @param {Library|Collection} libraryOrCollection
 	 * @param {Node<menupopup>} elem parent element
-	 * @param {function} clickAction function to execute on clicking the menuitem.
+	 * @param {Zotero.Library|Zotero.Collection} Currently selected item (displays as checked)
```
Comment (dstillman) [inline chrome/content/zotero/xpcom/utilities_internal.js:1698]: Missing "currentTarget"

**Review by dstillman — CHANGES_REQUESTED** [review]

**Pull request merged**